File: 1428-jump-game-iii/jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        const int n = arr.size();

        vector<short> dp(n, -1);
        function<bool(int)> dfs = [&](int i) {
            if (i < 0 || i >= n) {
                return false;
            }
            if (dp[i] != -1) {
                return dp[i] ? true : false;
            }
            if (arr[i] == 0) {
                dp[i] = 1;
                return true;
            }

            dp[i] = 0;
            if (dfs(i + arr[i]) || dfs(i - arr[i])) {
                dp[i] = 1;
            } else {
                dp[i] = 0;
            }

            return dp[i] ? true : false;
        };

        return dfs(start);
    }
};

File: 1428-jump-game-iii/jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        N = len(arr)
        dp = [False for _ in range(N)]
        
        dp[start] = True
        q = deque([start])
        while q:
            curr = q.popleft()
            if arr[curr] == 0:
                return True
            dp[curr] = True

            if curr - arr[curr] >= 0 and curr - arr[curr] < N and not dp[curr - arr[curr]]:
                dp[curr - arr[curr]] = True
                q.append(curr - arr[curr])
            if curr + arr[curr] >= 0 and curr + arr[curr] < N and not dp[curr + arr[curr]]:
                dp[curr + arr[curr]] = True
                q.append(curr + arr[curr])

        return False
